File: bufpay_api_demo.php
#!/usr/bin/env php

<?php

    #注意: 使用之前先到 bufpay 后台上传微信、支付宝App生成的收款二维码

    $price = $_GET('price'); # 获取充值金额
    $order_id = '101';       # 自己创建的本地订单号
    $order_uid = 'hello@hello.com';  # 订单对应的用户id
    $name = 'vip year fee';  # 订单商品名称
    $pay_type = 'wechat';    # 付款方式
    $notify_url = 'http://xxxx.com/notify_url.php';   # 回调通知地址
    $return_url = 'http://xxxx.com/return_url.php';   # 支付成功页面跳转地址

    $secret = 'your app secret';     # app secret, 在个人中心配置页面查看
    $api_url = 'https://bufpay.com/api/pay/XXX';   # 付款请求接口，在个人中心配置页面查看

    function sign($data_arr) {
        return md5(join('',$data_arr));
    };

    $sign = sign(array($name, $pay_type, $price, $order_id, $order_uid, $notify_url, $return_url, $secret));


echo '<html>
      <head><title>redirect...</title></head>
      <body>
          <form id="post_data" action="'.$api_url.'" method="post">
              <input type="hidden" name="name" value="'.$name.'"/>
              <input type="hidden" name="pay_type" value="'.$pay_type.'"/>
              <input type="hidden" name="price" value="'.$price.'"/>
              <input type="hidden" name="order_id" value="'.$order_id.'"/>
              <input type="hidden" name="order_uid" value="'.$order_uid.'"/>
              <input type="hidden" name="notify_url" value="'.$notify_url.'"/>
              <input type="hidden" name="return_url" value="'.$return_url.'"/>
              <input type="hidden" name="sign" value="'.$sign.'"/>
          </form>
          <script>document.getElementById("post_data").submit();</script>
      </body>
      </html>';
?>

File: bufpay_api_demo.py
#!/usr/bin/env python
#coding: utf8

import requests
import json
import hashlib


def sign(*p):
    return hashlib.md5(u''.join(p).encode('utf8')).hexdigest().lower()


def pay():
    #如果是 web 应用，这里先获取用户充值参数，然后创建本地订单，然后把构造好的参数 post 到 bufpay
    #如果希望用 bufpay 的支付页面，这里可以返回一个构造好的表单给用户（如下），在这个表单里面 post 到 bufpay

    #<!DOCTYPE html>
    #<html>
    #    <head><title>redirect...</title></head>
    #    <body>
    #        <form id="post_data" action="{{bufpay_api}}" method="post">
    #            {% for k, v in data.items() %}
    #            <input type="hidden" name="{{k}}" value="{{v}}"/>
    #            {% endfor%}
    #        </form>
    #        <script>document.getElementById("post_data").submit();</script>
    #    </body>
    #</html>

    pay_data = {
        'name': u'内容订阅一年期',
        'pay_type': 'wechat',
        'price': '50.00',
        'order_id': '888',
        'order_uid': 'hi@sideidea.com',
        'notify_url': 'http://sideidea.com/bufpay_notify',
        'return_url': 'http://sideidea.com'
    }
    pay_data['sign'] = sign(
        pay_data['name'],
        pay_data['pay_type'],
        pay_data['price'],
        pay_data['order_id'],
        pay_data['order_uid'],
        pay_data['notify_url'],
        pay_data['return_url'],
        'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'  #app secret, 这里需要修改为自己的
    )
    resp = requests.post('https://bufpay.com/api/pay/2?format=json', data=pay_data)
    return json.loads(resp.text)


def query(aoid):
    resp = requests.get('https://bufpay.com/api/query/' + aoid)
    return json.loads(resp.text)


if __name__ == '__main__':
    resp = pay()
    print resp
    print query(resp['aoid'])
